Collect info from multiple interfaces

Hello everyone.

I am currently trying to set up yanic for the Freifunk Aachen community.

In this community we have divided the network into multiple segments to keep the broadcast domain size managable etc. The monitoring server (which does among other tasks nodeinfo collection) has a bridge interface for each of our segments. As the seperation of segments is only a technical and not a social divide, we want to display all nodes from all segments on our map. Until now we used hopglass-server which is able to listen to multiple interfaces at the same time. This is not possible with yanic (Or is it? If so: Problem solved, all that remains is maybe a documentation fix?)

We cannot just bridge all these bridge interfaces together, as this would break stuff on the network side (hazy on the details but I think batman-adv bridge-loop-avoidance gets confused there).

I see the following solutions:

* Run one instance of Yanic on each of these bridges. A downstream script can then merge together the resulting nodes.json files to have one big file we can feed to our map.
** Downside: Having one config file per segment is kind of annoying when one wants to do global changes.
** Downside: Yet another work step when adding new segments (setting up a new yanic instance instead of changing e.g. a flag).
** Downside: One cannot use convenience features like the built-in webserver etc. anymore
** Upside: Would work right now without code changes. Having the merge script upstream would be kind of neat still.

* Add a configuration option similar to the one in hopglass-server where yanic can collect info from multiple interfaces at the same time.
** Downside: No idea how hard this is to implement.
** Upside: No additional data-plumbing required, everything should just work for the user.
** Upside: Network operators can change segments spontaneously or maybe even in an automated fashion.

If you have other ideas, I am open for input.

Thanks for your consideration!

I get the following panic:

```
$ yanic serve --config /etc/yanic.conf
2017/10/17 21:32:44 nodes.go:192: loaded 0 nodes
2017/10/17 21:32:44 serve.go:50: delaying 15.8 seconds
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/FreifunkBremen/yanic/respond.(*Collector).listenUDP(0xc420184000, 0xc4200157b0, 0x9)
	/opt/go/src/github.com/FreifunkBremen/yanic/respond/collector.go:73 +0x26d
github.com/FreifunkBremen/yanic/respond.NewCollector(0x9fec80, 0xc42010c000, 0xc42010c090, 0xc420016e00, 0x8, 0x8, 0x0, 0x32)
	/opt/go/src/github.com/FreifunkBremen/yanic/respond/collector.go:41 +0x12d
github.com/FreifunkBremen/yanic/cmd.glob..func3(0xa29a00, 0xc4200d0880, 0x0, 0x2)
	/opt/go/src/github.com/FreifunkBremen/yanic/cmd/serve.go:54 +0x1b8
github.com/spf13/cobra.(*Command).execute(0xa29a00, 0xc4200d0840, 0x2, 0x2, 0xa29a00, 0xc4200d0840)
	/opt/go/src/github.com/spf13/cobra/command.go:702 +0x2bd
github.com/spf13/cobra.(*Command).ExecuteC(0xa297e0, 0xc420074058, 0x0, 0x83ff14)
	/opt/go/src/github.com/spf13/cobra/command.go:783 +0x349
github.com/spf13/cobra.(*Command).Execute(0xa297e0, 0x8489ea, 0x1a)
	/opt/go/src/github.com/spf13/cobra/command.go:736 +0x2b
github.com/FreifunkBremen/yanic/cmd.Execute()
	/opt/go/src/github.com/FreifunkBremen/yanic/cmd/root.go:25 +0x31
main.main()
	/opt/go/src/github.com/FreifunkBremen/yanic/main.go:6 +0x20
```

Config file is as follows:
```

# Send respondd request to update information
[respondd]
enable           = true
# Delay startup until a multiple of the period since zero time
synchronize      = "1m"
# how oftern request per multicast
collect_interval = "1m"
# on which interface
interfaces        = [ "br-aachen", "br-seg-01", "br-seg-02", "br-seg-03", "br-seg-04", "br-seg-05", "br-seg-06", "br-seg-07" ]
#interface = "br-aachen"
#interface = "br-seg-01" 
# define a port to listen
# (no or 0 would choose at port at his own)
#port = 10001


# A little build-in webserver, which statically serves a directory.
# This is useful for testing purposes or for a little standalone installation.
[webserver]
enable  = false
bind    = "127.0.0.1:8080"
webroot = "/var/www/html/meshviewer"



[nodes]
enable         = true
# state-version of nodes.json to store cached data,
#  these is the directly collected respondd data
state_path     = "/var/lib/yanic/state.json"

# Export nodes and graph periodically
save_interval = "5s"

# Set node to offline if not seen within this period
offline_after = "10m"

# Prune offline nodes after a time of inactivity
prune_after = "7d"


[meshviewer]
# structur of nodes.json, which to support
# version 1 is to support legacy meshviewer (which are in master branch)
#    i.e. https://github.com/ffnord/meshviewer/tree/master
# version 2 is to support new version of meshviewer (which are in legacy develop branch or newer)
#    i.e. https://github.com/ffnord/meshviewer/tree/dev
#         https://github.com/ffrgb/meshviewer/tree/develop
version  = 2
# path where to store nodes.json
nodes_path     = "/var/lib/yanic/output/nodes.json"
# path where to store graph.json
graph_path     = "/var/lib/yanic/output/graph.json"

[database]
# cleaning data of measurement node,
#   which are older than 7d
delete_after = "7d"
#   how often run the cleaning
delete_interval = "1h"

# Save collected data to InfluxDB
# there would be the following measurments:
#  node: store node spezific data i.e. clients memory, airtime
#  global: store global data, i.e. count of clients and nodes
#  firmware: store count of nodes tagged with firmware
#  model: store count of nodes tagged with hardware model
[[database.connection.influxdb]]
enable   = false
address  = "http://localhost:8086"
database = "ffhb"
username = ""
password = ""
# tagging of the data are optional
# be carefull tags by system would overright config
[database.connection.influxdb.tags]
site = "ffhb01"
system = "testing"

[[database.connection.logging]]
enable   = false
path     = "/var/log/yanic.log"

[[database.connection.graphite]]
enable   = false
address  = "localhost:2003"
prefix   = "freifunk"

```

--- a/cmd/query.go
+++ b/cmd/query.go
@@ -26,7 +26,7 @@ var queryCmd = &cobra.Command{
 
 		nodes := runtime.NewNodes(&runtime.Config{})
 
-		collector := respond.NewCollector(nil, nodes, iface, 0)
+		collector := respond.NewCollector(nil, nodes, []string{iface}, 0)
 		defer collector.Close()
 		collector.SendPacket(dstAddress)
 

--- a/cmd/serve.go
+++ b/cmd/serve.go
@@ -51,7 +51,7 @@ var serveCmd = &cobra.Command{
 				time.Sleep(delay)
 			}
 
-			collector = respond.NewCollector(connections, nodes, config.Respondd.Interface, config.Respondd.Port)
+			collector = respond.NewCollector(connections, nodes, config.Respondd.Interfaces, config.Respondd.Port)
 			collector.Start(config.Respondd.CollectInterval.Duration)
 			defer collector.Close()
 		}

--- a/config_example.toml
+++ b/config_example.toml
@@ -6,7 +6,7 @@ synchronize      = "1m"
 # how oftern request per multicast
 collect_interval = "1m"
 # on which interface
-interface        = "eth0"
+interfaces       = ["eth0"]
 # define a port to listen
 # (no or 0 would choose at port at his own)
 #port = 10001

--- a/respond/collector.go
+++ b/respond/collector.go
@@ -17,17 +17,43 @@ import (
 
 // Collector for a specificle respond messages
 type Collector struct {
-	connection *net.UDPConn   // UDP socket
-	queue      chan *Response // received responses
-	iface      string
-	db         database.Connection
-	nodes      *runtime.Nodes
-	interval   time.Duration // Interval for multicast packets
-	stop       chan interface{}
+	connections []*net.UDPConn          // UDP sockets
+	ifaceToConn map[string]*net.UDPConn // map from interface name to UDP socket
+
+	queue    chan *Response // received responses
+	db       database.Connection
+	nodes    *runtime.Nodes
+	interval time.Duration // Interval for multicast packets
+	stop     chan interface{}
 }
 
 // NewCollector creates a Collector struct
-func NewCollector(db database.Connection, nodes *runtime.Nodes, iface string, port int) *Collector {
+func NewCollector(db database.Connection, nodes *runtime.Nodes, ifaces []string, port int) *Collector {
+
+	coll := &Collector{
+		db:    db,
+		nodes: nodes,
+		queue: make(chan *Response, 400),
+		stop:  make(chan interface{}),
+	}
+
+	for _, iface := range ifaces {
+		coll.listenUDP(iface)
+	}
+
+	go coll.parser()
+
+	if coll.db != nil {
+		go coll.globalStatsWorker()
+	}
+
+	return coll
+}
+
+func (coll *Collector) listenUDP(iface string) {
+	if _, found := coll.ifaceToConn[iface]; found {
+		log.Panicf("can not listen twice on %s", iface)
+	}
 	linkLocalAddr, err := getLinkLocalAddr(iface)
 	if err != nil {
 		log.Panic(err)
@@ -44,23 +70,11 @@ func NewCollector(db database.Connection, nodes *runtime.Nodes, iface string, po
 	}
 	conn.SetReadBuffer(maxDataGramSize)
 
-	collector := &Collector{
-		connection: conn,
-		db:         db,
-		nodes:      nodes,
-		iface:      iface,
-		queue:      make(chan *Response, 400),
-		stop:       make(chan interface{}),
-	}
+	coll.ifaceToConn[iface] = conn
+	coll.connections = append(coll.connections, conn)
 
-	go collector.receiver()
-	go collector.parser()
-
-	if collector.db != nil {
-		go collector.globalStatsWorker()
-	}
-
-	return collector
+	// Start receiver
+	go coll.receiver(conn)
 }
 
 // Returns the first link local unicast address for the given interface name
@@ -102,7 +116,9 @@ func (coll *Collector) Start(interval time.Duration) {
 // Close Collector
 func (coll *Collector) Close() {
 	close(coll.stop)
-	coll.connection.Close()
+	for _, conn := range coll.connections {
+		conn.Close()
+	}
 	close(coll.queue)
 }
 
@@ -116,8 +132,10 @@ func (coll *Collector) sendOnce() {
 }
 
 func (coll *Collector) sendMulticast() {
-	log.Println("sending multicast")
-	coll.SendPacket(net.ParseIP(multiCastGroup))
+	log.Println("sending multicasts")
+	for _, conn := range coll.connections {
+		coll.sendPacket(conn, multiCastGroup)
+	}
 }
 
 // Send unicast packets to nodes that did not answer the multicast
@@ -132,20 +150,30 @@ func (coll *Collector) sendUnicasts(seenBefore jsontime.Time) {
 	// Send unicast packets
 	log.Printf("sending unicast to %d nodes", len(nodes))
 	for _, node := range nodes {
-		coll.SendPacket(node.Address)
+		conn := coll.ifaceToConn[node.Address.Zone]
+		if conn == nil {
+			log.Printf("unable to find connection for %s", node.Address.Zone)
+			continue
+		}
+		coll.sendPacket(conn, node.Address.IP)
 		time.Sleep(10 * time.Millisecond)
 	}
 }
 
-// SendPacket sends a UDP request to the given unicast or multicast address
-func (coll *Collector) SendPacket(address net.IP) {
+// SendPacket sends a UDP request to the given unicast or multicast address on the first UDP socket
+func (coll *Collector) SendPacket(destination net.IP) {
+	coll.sendPacket(coll.connections[0], destination)
+}
+
+// sendPacket sends a UDP request to the given unicast or multicast address on the given UDP socket
+func (coll *Collector) sendPacket(conn *net.UDPConn, destination net.IP) {
 	addr := net.UDPAddr{
-		IP:   address,
+		IP:   destination,
 		Port: port,
-		Zone: coll.iface,
+		Zone: conn.LocalAddr().(*net.UDPAddr).Zone,
 	}
 
-	if _, err := coll.connection.WriteToUDP([]byte("GET nodeinfo statistics neighbours"), &addr); err != nil {
+	if _, err := conn.WriteToUDP([]byte("GET nodeinfo statistics neighbours"), &addr); err != nil {
 		log.Println("WriteToUDP failed:", err)
 	}
 }
@@ -187,7 +215,7 @@ func (res *Response) parse() (*data.ResponseData, error) {
 	return rdata, err
 }
 
-func (coll *Collector) saveResponse(addr net.UDPAddr, res *data.ResponseData) {
+func (coll *Collector) saveResponse(addr *net.UDPAddr, res *data.ResponseData) {
 	// Search for NodeID
 	var nodeID string
 	if val := res.NodeInfo; val != nil {
@@ -217,7 +245,7 @@ func (coll *Collector) saveResponse(addr net.UDPAddr, res *data.ResponseData) {
 
 	// Process the data and update IP address
 	node := coll.nodes.Update(nodeID, res)
-	node.Address = addr.IP
+	node.Address = addr
 
 	// Store statistics in database
 	if db := coll.db; db != nil {
@@ -232,10 +260,10 @@ func (coll *Collector) saveResponse(addr net.UDPAddr, res *data.ResponseData) {
 	}
 }
 
-func (coll *Collector) receiver() {
+func (coll *Collector) receiver(conn *net.UDPConn) {
 	buf := make([]byte, maxDataGramSize)
 	for {
-		n, src, err := coll.connection.ReadFromUDP(buf)
+		n, src, err := conn.ReadFromUDP(buf)
 
 		if err != nil {
 			log.Println("ReadFromUDP failed:", err)
@@ -246,7 +274,7 @@ func (coll *Collector) receiver() {
 		copy(raw, buf)
 
 		coll.queue <- &Response{
-			Address: *src,
+			Address: src,
 			Raw:     raw,
 		}
 	}

--- a/respond/collector.go
+++ b/respond/collector.go
@@ -31,10 +31,11 @@ type Collector struct {
 func NewCollector(db database.Connection, nodes *runtime.Nodes, ifaces []string, port int) *Collector {
 
 	coll := &Collector{
-		db:    db,
-		nodes: nodes,
-		queue: make(chan *Response, 400),
-		stop:  make(chan interface{}),
+		db:          db,
+		nodes:       nodes,
+		queue:       make(chan *Response, 400),
+		stop:        make(chan interface{}),
+		ifaceToConn: make(map[string]*net.UDPConn),
 	}
 
 	for _, iface := range ifaces {

--- a/respond/collector_test.go
+++ b/respond/collector_test.go
@@ -3,9 +3,20 @@ package respond
 import (
 	"io/ioutil"
 	"testing"
+	"time"
 
+	"github.com/FreifunkBremen/yanic/runtime"
 	"github.com/stretchr/testify/assert"
 )
+
+func TestCollector(t *testing.T) {
+	nodes := runtime.NewNodes(&runtime.Config{})
+
+	collector := NewCollector(nil, nodes, []string{}, 10001)
+	collector.Start(time.Millisecond)
+	time.Sleep(time.Millisecond * 10)
+	collector.Close()
+}
 
 func TestParse(t *testing.T) {
 	assert := assert.New(t)

--- a/respond/respond.go
+++ b/respond/respond.go
@@ -4,9 +4,10 @@ import (
 	"net"
 )
 
+// default multicast group used by announced
+var multiCastGroup = net.ParseIP("ff02:0:0:0:0:0:2:1001")
+
 const (
-	// default multicast group used by announced
-	multiCastGroup = "ff02:0:0:0:0:0:2:1001"
 
 	// default udp port used by announced
 	port = 1001
@@ -15,8 +16,8 @@ const (
 	maxDataGramSize = 8192
 )
 
-//Response of the respond request
+// Response of the respond request
 type Response struct {
-	Address net.UDPAddr
+	Address *net.UDPAddr
 	Raw     []byte
 }

--- a/runtime/config.go
+++ b/runtime/config.go
@@ -11,7 +11,7 @@ type Config struct {
 	Respondd struct {
 		Enable          bool     `toml:"enable"`
 		Synchronize     Duration `toml:"synchronize"`
-		Interface       string   `toml:"interface"`
+		Interfaces      []string `toml:"interfaces"`
 		Port            int      `toml:"port"`
 		CollectInterval Duration `toml:"collect_interval"`
 	}

--- a/runtime/config_test.go
+++ b/runtime/config_test.go
@@ -15,7 +15,7 @@ func TestReadConfig(t *testing.T) {
 	assert.NotNil(config)
 
 	assert.True(config.Respondd.Enable)
-	assert.Equal("eth0", config.Respondd.Interface)
+	assert.Equal([]string{"eth0"}, config.Respondd.Interfaces)
 	assert.Equal(time.Minute, config.Respondd.CollectInterval.Duration)
 
 	assert.Equal(2, config.Meshviewer.Version)

--- a/runtime/node.go
+++ b/runtime/node.go
@@ -9,7 +9,7 @@ import (
 
 // Node struct
 type Node struct {
-	Address    net.IP           `json:"address"` // the last known IP address
+	Address    *net.UDPAddr     `json:"-"` // the last known address
 	Firstseen  jsontime.Time    `json:"firstseen"`
 	Lastseen   jsontime.Time    `json:"lastseen"`
 	Online     bool             `json:"online"`

